library/core/src/slice/iter.rs

```diff
@@ -157,6 +160,31 @@ impl<T> AsRef<[T]> for Iter<'_, T> {
     }
 }
 
+#[unstable(feature = "ub_checks", issue = "none")]
+impl<T> Invariant for Iter<'_, T> {
+    /// An iterator can be safely used if its pointer can be read for its current length.
+    ///
+    /// If the type is a ZST or the encoded length is `0`, the only safety requirement is that
+    /// its pointer is aligned (since zero-size access is always safe for aligned pointers).
+    ///
+    /// For other cases, we need to ensure that it is safe to read the memory between
+    /// `self.ptr` and `self.end_or_len`.
+    fn is_safe(&self) -> bool {
+        let ty_size = crate::mem::size_of::<T>();
+        let distance = self.ptr.addr().get().abs_diff(self.end_or_len as usize);
```

Comment:
Because this would overflow if `ptr` is greater than `end_or_len` which we only check below. So we either move this statement down, or we use `abs_diff` or `wrapping_sub`.

Reply:
In what scenarios would `ptr` be greater than `end_or_len`?

Reply:
Talked offline about this. This can happen in the case of ZST. To avoid confusion, I'll add a comment.